Automattic/woocommerce-payments
# Associate saved card with the customer

Fixes #965

#### Changes proposed in this Pull Request

* Add customer ID to token's meta when creating it
* Do not display tokens that do not belong to current customer
* Add customer information to old tokens based on account mode and created customers
* Do not delete tokens that belong to the inactive customer

#### Testing instructions



1. **Test accounts**
    1.1. Using a test account, add a card while using `trunk`
    1.2. Apply this PR and add a new card
    1.3. Assert that both tokens are displayed during checkout
    1.4. Assert that both tokens have a `_wcpay_customer_id` meta added to them: The token created with `trunk` should've been migrated, as it shouldn't have a customer ID before this PR is applied
2. **Live accounts in test mode without live customer**
    3.1. Using a live account in test mode, add a card while using `trunk`
    3.2. Apply this PR and add a new card
    3.3. Assert that both tokens are displayed during checkout
    3.4. Assert that both tokens have a `_wcpay_customer_id` meta added to them: The token created with `trunk` should've been migrated, as it shouldn't have a customer ID before this PR is applied
3. **Live accounts in test mode with live customer**
    3.1. Ensure your store has a live-mode customer before switching test mode on
    3.2. Using the live account in test mode, add a card while using `trunk`
    3.3. Apply this PR and add a new card
    3.4. Assert that only the new token is displayed during checkout
    3.5. Assert that both tokens have a `_wcpay_customer_id` meta added to them: The token created with `trunk` should've been migrated to the customer from live mode
    3.6. Switch back to live mode, you should see the token from `3.2` in the checkout (even though it won't work)
4. **Live accounts in live mode**
    4.1. Using a live account in live mode, add a card while using `trunk`
    4.2. Apply this PR and add a new card
    4.3. Assert that both tokens are displayed during checkout
    4.4. Assert that both tokens have a `_wcpay_customer_id` meta added to them: The token created with `trunk` should've been migrated, as it shouldn't have a customer ID before this PR is applied
    4.5. Switch the store to test mode and add a new card
    4.6. Assert that a new customer was created under the new test key
    4.7. Assert that only the test token is displayed during checkout
    4.8. Assert that the new token has the test customer in the `_wcpay_customer_id` meta

-------------------
